lib/dfProduct_build.php

```diff
@@ -0,0 +1,371 @@
+            $this->products
+        );
+
+        return $products;
+    }
+
+    private function buildProduct($product)
+    {
+        $p = array();
+
+        $p['id'] = $this->getId($product);
+        $p['title'] = dfTools::cleanString($product['name']);
+        $p['link'] = $this->getLink($product);
+        $p['description'] = dfTools::cleanString($product['description_short']);
+        $p['alternate_description'] = dfTools::cleanString($product['description']);
+        $p['meta_keywords'] = dfTools::cleanString($product['meta_keywords']);
+        $p['meta_title'] = dfTools::cleanString($product['meta_title']);
+        $p['meta_description'] = dfTools::cleanString($product['meta_description']);
+        $p['image_link'] = $this->getImageLink($product);
+        $p['categories'] = dfTools::getCategoriesForProductIdAndLanguage(
+            $product['id_product'],
+            $this->id_lang,
+            $this->id_shop,
+            false
+        );
+        $p['availability'] = $this->getAvailability($product);
+        $p['brand'] = dfTools::cleanString($product['manufacturer']);
+        $p['mpn'] = dfTools::cleanString($product['mpn']);
+        $p['ean13'] = dfTools::cleanString($product['ean13']);
+        $p['upc'] = dfTools::cleanString($product['upc']);
+        $p['reference'] = dfTools::cleanString($product['reference']);
+        $p['supplier_reference'] = dfTools::cleanString($product['supplier_reference']);
+        $p['extra_title_1'] = dfTools::cleanReferences($p['title']);
+        $p['extra_title_2'] = dfTools::splitReferences($p['title']);
+        $p['tags'] = dfTools::cleanString($product['tags']);
+
+        if (dfTools::versionGte('1.7.0.0')) {
+            $p['isbn'] = dfTools::cleanString($product['isbn']);
+        }
+
+        if ($this->display_prices) {
+            $p['price'] = $this->getPrice($product);
+            $p['sale_price'] = $this->getPrice($product, true);
+        }
+
+        if ($this->show_product_features) {
+            $p['attributes'] = $this->getFeatures($product);
+        }
+
+        if ($this->product_variations) {
+            $p['item_group_id'] = $this->getItemGroupId($product);
+
+            $p['variation_reference'] = $product['variation_reference'];
+            $p['variation_supplier_reference'] = $product['variation_supplier_reference'];
+            $p['variation_mpn'] = $product['variation_mpn'];
+            $p['variation_ean13'] = $product['variation_ean13'];
+            $p['variation_upc'] = $product['variation_upc'];
+            $p['df_group_leader'] = (!is_null($product['df_group_leader']) ? true : false);
+
+            $attributes = $this->getAttributes($product);
+
+            $p = array_merge($p, $attributes);
+        }
+
+        return $p;
+    }
+
+    private function getId($product)
+    {
+        if ($this->product_variations) {
+            if (isset($product['id_product_attribute']) && (int)$product['id_product_attribute'] > 0) {
```

Comment:
this code is repeated in various functions, maybe we should decouple in another function

Reply:
replaced by a function to avoid repeating code